### DISCUSSION TITLE
Custom SQL Queries

### DISCUSSION THREAD

**Discussion by iamvaliyev** [directus/directus#3312] [Archive]
I want write custom sql query. For this there is a ready-made solution ?

**Comment by rijkvanzanten** [comment]
You'll have to be more specific in where/when/why/how etc. Directus runs on multiple SQL vendors, so there's no native way to just "run" a query (that would also be wildly unsafe).

You can write a custom endpoint or hook if you truly need to run direct SQL queries, but I'd advice against it

**Reply by iamvaliyev** [reply]
> You'll have to be more specific in where/when/why/how etc. Directus runs on multiple SQL vendors, so there's no native way to just "run" a query (that would also be wildly unsafe).
> 
> You can write a custom endpoint or hook if you truly need to run direct SQL queries, but I'd advice against it

I write custom endpoint and inside I want to write my query. Database is postgresql

**Reply by mehmetuken** [reply]
https://github.com/directus/directus/blob/main/docs/guides/extensions/api-endpoints.md see this document and expose database this project using knex sql query builder read raw sql knex document
http://knexjs.org/#Raw-Queries

example.

```js
module.exports = function registerEndpoint(router, {exceptions, database }) {
	const { ServiceUnavailableException } = exceptions;

	router.get('/', (req, res) => {
		database.raw('SELECT first_name FROM directus_users')
			.then((results) => res.json(results))
			.catch((error) => {
				throw new ServiceUnavailableException(error.message);
			});
	});
};
```

**Reply by iamvaliyev** [reply]
> https://github.com/directus/directus/blob/main/docs/guides/extensions/api-endpoints.md see this document and expose database this project using knex sql query builder read raw sql knex document
> http://knexjs.org/#Raw-Queries
> 
> example.
> 
> ```js
> module.exports = function registerEndpoint(router, {exceptions, database }) {
> 	const { ServiceUnavailableException } = exceptions;
> 
> 	router.get('/', (req, res) => {
> 		database.raw('SELECT first_name FROM directus_users')
> 			.then((results) => res.json(results))
> 			.catch((error) => {
> 				throw new ServiceUnavailableException(error.message);
> 			});
> 	});
> };
> ```

thanks 🙏🏻🙏🏻🙏🏻🙏🏻🙏🏻

**Reply by flying-sheep** [reply]
Here’s the updated docs link: https://docs.directus.io/extensions/endpoints/

**Reply by ddelange** [reply]
Here is a bit more complete typescript example that runs on Directus v11:

Note: after following [Getting Started](https://directus.io/docs/guides/extensions/quickstart), I still had to `pnpm add @directus/errors` before `pnpm run dev`.

```ts
import { createError } from '@directus/errors';
import { defineEndpoint, ApiExtensionContext } from '@directus/extensions-sdk';
import type { Knex } from 'knex';
import type { Request, Response, Router, NextFunction } from 'express';

const ForbiddenError = createError('FORBIDDEN', "You don't have permissions to see this.", 403);

export default defineEndpoint((router: Router, context: ApiExtensionContext) => {
	/**
	 * GET /
	 *
	 * Search companies by terms in their Wikipedia extract.
	 *
	 * Query Parameters:
	 * - term: string or array - Search term(s) to filter by. Multiple terms are combined with AND.
	 *   Example: ?term=oil&term=gas (finds companies with both "oil" AND "gas" in extract)
	 * - limit: number - Maximum results to return (default: 10)
	 * - order_by: string or array - Field(s) to sort by, optionally with direction (default: "name")
	 *   Examples:
	 *   - ?order_by=name (ascending by default)
	 *   - ?order_by=-name (descending with minus prefix)
	 *   - ?order_by=industry&order_by=-name (multiple sorts applied in order)
	 */
	router.get('/', async (req: Request, res: Response, next: NextFunction) => {
		try {
			let accountability = req.accountability;  // Accountability ref https://github.com/directus/directus/blob/v11.12.0/packages/types/src/accountability.ts
			if (accountability.role === null) {
				throw new ForbiddenError();
			}

			let database: Knex = context.database;  // ApiExtensionContext ref https://github.com/directus/directus/blob/v11.12.0/packages/types/src/extensions/api-extension-context.ts

			// Parse query parameters
			const limit = parseInt(req.query.limit as string) || 10;
			const orderByParam = req.query.order_by || 'name';
			const orderByArray = Array.isArray(orderByParam) ? orderByParam : [orderByParam];
			const termParam = req.query.term;
			const termArray = Array.isArray(termParam) ? termParam : (termParam ? [termParam] : []);

			let query = database('companies').select('name', 'wikipedia_extract').limit(limit);

			// Apply ordering
			orderByArray.forEach((field: string) => {
				if (field.startsWith('-')) {
					query = query.orderBy(field.substring(1), 'desc');
				} else {
					query = query.orderBy(field, 'asc');
				}
			});

			// Apply AND conditions for each search term
			termArray.forEach((term: string) => {
				query = query.whereILike('wikipedia_extract', `%${term}%`);
			});

			const results = await query;
			res.json(results);
		} catch (error) {
			next(error);
		}
	});
});
```
